### PR TITLE
AVRO-4136: [c] json encoding of byte[] containing 0x00

### PR DESCRIPTION
**Pull request by steven-aerts**
Prevent the fixed and bytes type to be preliminary cut off when they are encoded into the json encoding.

## Verifying this change

This change added tests and can be verified as follows: We updated the existing `test_avro_data` tests to validate this corner case.

## Documentation

- Does this pull request introduce a new feature? no

### CODE REVIEW

File: lang/c/src/value-json.c [hunk 1]
```diff
@@ -29,7 +29,7 @@
 #include "jansson.h"
 
 /*
- * Converts a binary buffer into a NUL-terminated JSON UTF-8 string.
+ * Converts a binary buffer into a JSON UTF-8 string which is not NULL terminated!
```
Comment (KalleOlaviNiemitalo) [inline lang/c/src/value-json.c]: This comment is a bit misleading.   `encode_utf8_bytes` does not encode control characters as `\u0001` etc. like JSON requires, so the result isn't really a "JSON UTF-8 string".

I'd prefer "null-terminated" to "NULL-terminated" because the `NULL` macro in \<stddef.h\> is intended as a pointer rather than a character.

File: lang/c/src/value-json.c [hunk 2]
```diff
@@ -29,7 +29,8 @@
 #include "jansson.h"
 
 /*
- * Converts a binary buffer into a NUL-terminated JSON UTF-8 string.
+ * Converts a binary buffer into a JSON UTF-8 string which is NOT
+ * terminated with a null byte ('\0).
```
Comment (martin-g) [inline lang/c/src/value-json.c]:
```suggestion
 * terminated with a null byte ('\0').
```

**Review by martin-g — APPROVED** [review]
LGTM!

How about adding some extra tests for:
* char all_nulls[] = {0x00, 0x00, 0x00};
* char starts_with_null[] = {0x00, 0xDE, 0xAD};
* char ends_with_null[] = {0xDE, 0xAD, 0x00};

File: lang/c/src/value-json.c
Comment (martin-g) [inline lang/c/src/value-json.c:64]: I'm not sure whether this is a real use case:
Is it possible someone to use a Bytes schema with an empty value, i.e. zero bytes ?!
In this case `utf8_len` would be `0` and the malloc would return NULL and fail below with ENOMEM.

File: lang/c/src/value-json.c [hunk 3]
```diff
@@ -127,7 +126,7 @@ avro_value_to_json_t(const avro_value_t *value)
 				return NULL;
 			}
 
-			json_t  *result = json_string_nocheck((const char *) encoded);
+			json_t  *result = json_stringn_nocheck((const char *) encoded, encoded_size);
```
Comment (martin-g) [inline lang/c/src/value-json.c:129]: https://jansson.readthedocs.io/en/latest/apiref.html says that json_stringn_nocheck was added in jansson 2.7

But we require 2.3+ at https://github.com/apache/avro/blob/a60a68f2c335aa1af943481b096c9941fd2bceda/lang/c/CMakeLists.txt#L182
We need to bump it to 2.7+

File: lang/c/src/value-json.c
Comment (steven-aerts) [reply lang/c/src/value-json.c:64]: An empty byte array should be possible in avro.
When I tried making a test, I see it failing at other places making similar mistakes.
Propose to create a separate patch with this test.

File: lang/c/src/value-json.c (hunk 3 above)
Comment (steven-aerts) [reply lang/c/src/value-json.c:129]: Nice catch

File: lang/c/src/value-json.c
Comment (martin-g) [reply lang/c/src/value-json.c:64]: Agreed! Thank you!

**Comment by martin-g** [comment]
Thank you, @steven-aerts !

File: lang/c/src/value-json.c
Comment (steven-aerts) [reply lang/c/src/value-json.c:64]: https://github.com/apache/avro/pull/3527
